pkg/controller/hpaWatcher.go

```diff
@@ -0,0 +1,200 @@
+				})
+
+				if err == nil {
+					queue.Add(string(out))
+				}
+			}
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+
+			if err == nil {
+				out, err := json.Marshal(hpaEvent{
+					EventName:  receivers.UpdateEvent,
+					HpaName:    key,
+					NewHpaData: new.(*asv1.HorizontalPodAutoscaler),
+					OldHpaData: old.(*asv1.HorizontalPodAutoscaler),
+				})
+
+				if err == nil {
+					queue.Add(string(out))
+				}
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+
+			if err == nil {
+				out, err := json.Marshal(hpaEvent{
+					EventName:  receivers.DeleteEvent,
+					HpaName:    key,
+					NewHpaData: nil,
+					OldHpaData: obj.(*asv1.HorizontalPodAutoscaler),
+				})
+
+				if err == nil {
+					queue.Add(string(out))
+				}
+			}
+		},
+	}, cache.Indexers{})
+
+	hpaController = newController(queue, indexer, informer, hpaEventsHandler, "HorizontalPodAutoscaler")
+	return hpaController
+}
+
+// hpaEventsHandler is the business logic of the hpa controller.
+// In case an error happened, it has to simply return the error.
+func hpaEventsHandler(key string, indexer cache.Indexer) error {
+	event := hpaEvent{}
+	json.Unmarshal([]byte(key), &event)
+
+	var eventMessage string
+	var hpaAnnotations map[string]string
+	hpaWatchSlackUsersID := make([]string, 0)
+
+	if event.NewHpaData != nil {
+		hpaAnnotations = event.NewHpaData.GetObjectMeta().GetAnnotations()
+	}
+
+	eventReceivers := common.BuildEventReceiversList(hpaAnnotations)
+	log.Debug().Msg(fmt.Sprintf("found %d event receivers for HorizontalPodAutoscaler[%s]. receivers[%s]", len(eventReceivers), event.HpaName, strings.Join(eventReceivers, ",")))
+
+	switch event.EventName {
+	case "Add":
+		log.Debug().Msg(fmt.Sprintf("handling 'Add' event for HorizontalPodAutoscaler[%s]", event.HpaName))
+		eventMessage = fmt.Sprintf("New HorizontalPodAutoscaler resource [`%s`] added to `%s` cluster", event.HpaName, config.ClusterName())
+		log.Debug().Msg(eventMessage)
+
+	case "Delete":
+		log.Debug().Msg(fmt.Sprintf("handling 'Delete' event for HorizontalPodAutoscaler[%s]", event.HpaName))
+		eventMessage = fmt.Sprintf("HorizontalPodAutoscaler resource [`%s`] has deleted from `%s` cluster", event.HpaName, config.ClusterName())
+		log.Debug().Msg(eventMessage)
+
+	default:
+		// update hpa event
+		log.Debug().Msg(fmt.Sprintf("handling 'Update' event for HorizontalPodAutoscaler[%s]", event.HpaName))
+
+		var oldHPAStatus, newHPAStatus asv1.HorizontalPodAutoscalerStatus
+
+		if event.OldHpaData != nil && event.NewHpaData != nil {
+			oldHPAStatus = event.OldHpaData.Status
+			newHPAStatus = event.NewHpaData.Status
+		} else {
+			log.Warn().Msg(fmt.Sprintf("HorizontalPodAutoscaler [%s] old and/or new status is nil. unable to handle 'Update' event", event.EventName))
+			return nil
+		}
+
+		// new HPA event detected
+		if oldHPAStatus.CurrentReplicas == oldHPAStatus.CurrentReplicas {
```

Comment:
why wouldn't it be equal? or am I missing something?

Reply:
this is typo. fixed